### PR TITLE
test(spanner): limit special instance-config filtering to prod

### PR DESCRIPTION
**Pull request by devbww**
When selecting an `InstanceConfig` to use in the samples that create
a new `Instance`, we have some special filtering to avoid locations
where we run integration tests.  This filtering is specific to the
production Spanner service, and so should not be applied when using
an alternative like the staging service.

This allows the `spanner_create_database_with_encryption_key` sample
to run on the staging service, as its encryption key is tied to a
location whose `InstanceConfig` was previously discarded by the filter.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9075)

### CODE REVIEW

**Review by reinasalondra — APPROVED** [review]

File: google/cloud/spanner/samples/samples.cc [hunk 1]
```diff
@@ -3728,21 +3728,27 @@ void RunAllSlowInstanceTests(
...
+  auto const prod_service = google::cloud::internal::GetEnv(
+                                "GOOGLE_CLOUD_CPP_SPANNER_DEFAULT_ENDPOINT")
+                                .value_or("")
+                                .empty();
   auto const config_id =
       Basename(google::cloud::spanner_testing::PickInstanceConfig(
           google::cloud::Project(project_id), generator,
-          [](google::spanner::admin::instance::v1::InstanceConfig const&
-                 config) {
+          [prod_service](
```
Comment (coryan) [inline google/cloud/spanner/samples/samples.cc:3738]: nit: maybe time to move the functor out?

File: google/cloud/spanner/samples/samples.cc [hunk 2]
```diff
@@ -3728,21 +3728,27 @@ void RunAllSlowInstanceTests(
...
+                  config) {
             auto config_id = Basename(config.name());
             // Skip non-US configs. They are too slow.
-            if (absl::StartsWith(config_id, "regional-us-")) {
-              // Exclude US configs where we keep most test instances.
-              if (config_id != "regional-us-central1" &&
-                  config_id != "regional-us-east1") {
-                return true;
+            if (!absl::StartsWith(config_id, "regional-us-")) return false;
+            if (prod_service) {
```
Comment (coryan) [inline google/cloud/spanner/samples/samples.cc:3744]: maybe:

```cc
if (!prod_service) return true;
return config_id != "regional-us-central1" && config_id != "regional-us-east1";
```

**Review by coryan — APPROVED** [review]

File: google/cloud/spanner/samples/samples.cc (hunk 1 above)
Comment (devbww) [reply google/cloud/spanner/samples/samples.cc]: OK, done.

File: google/cloud/spanner/samples/samples.cc (hunk 2 above)
Comment (devbww) [reply google/cloud/spanner/samples/samples.cc]: Done.

**Review by coryan — APPROVED** [review]

**Pull request merged**